src/Processors/Transforms/CreatingSetsTransform.cpp

```diff
@@ -39,24 +40,63 @@ void CreatingSetsTransform::work()
     if (!is_initialized)
         init();
 
+    if (done_with_set && done_with_table)
+    {
+        finishConsume();
+        input.close();
+    }
+
     IAccumulatingTransform::work();
 }
 
 void CreatingSetsTransform::startSubquery()
 {
-    if (subquery.set)
-        LOG_TRACE(log, "Creating set.");
+    /// Lookup the set in the cache if we don't need to build table.
+    auto ctx = context.lock();
+    if (ctx && ctx->getPreparedSetsCache() && !subquery.table)
+    {
+        /// Try to find the set in the cache and wait for it to be built.
+        /// Retry if the set from cache fails to be built.
+        while (true)
+        {
+            auto from_cache = ctx->getPreparedSetsCache()->findOrPromiseToBuild(subquery.key);
+            if (from_cache.index() == 0)
+            {
+                promise_to_build = std::move(std::get<0>(from_cache));
+            }
+            else
+            {
+                LOG_TRACE(log, "Waiting for set to be build by another thread, key: {}", subquery.key);
+                SharedSet set_built_by_another_thread = std::move(std::get<1>(from_cache));
+                const SetPtr & ready_set = set_built_by_another_thread.get();
```

Comment:
Are we supposed to lock here and wait until builing is finished?

Reply:
Yes, we block here waiting the future value to be set.